src/orion/algo/registry.py

```diff
@@ -0,0 +1,147 @@
+
+    def register(self, trial: Trial) -> str:
+        """Register the given trial in the registry."""
+        trial_id = _get_id(trial)
+        if trial_id in self:
+            existing = self._trials[trial_id]
+            logger.debug(
+                f"Overwriting existing trial {existing} with new trial {trial}"
+            )
+        else:
+            logger.debug(
+                f"Registry {id(self)} Registering new trial {trial} ({len(self)} trials in total)"
+            )
+        trial_copy = copy.deepcopy(trial)
+        self._trials[trial_id] = trial_copy
+        return trial_id
+
+
+class RegistryMapping(Mapping[Trial, "list[Trial]"]):
+    """A map between the original and transformed registries.
+
+    This object is used in the `SpaceTransformAlgoWrapper` to check if a trial in the original space
+    has equivalent trials in the transformed space.
+
+    The goal is to make it so the algorithms don't have to care about the transforms/etc.
+    """
+
+    def __init__(self, original_registry: Registry, transformed_registry: Registry):
+        self.original_registry = original_registry
+        self.transformed_registry = transformed_registry
+        self._mapping: dict[str, set[str]] = defaultdict(set)
+
+    @property
+    def state_dict(self) -> dict:
+        return {
+            "original_registry": self.original_registry.state_dict,
+            "transformed_registry": self.transformed_registry.state_dict,
+            "_mapping": self._mapping,
+        }
+
+    def set_state(self, statedict: dict):
+        self.original_registry.set_state(statedict["original_registry"])
+        self.transformed_registry.set_state(statedict["transformed_registry"])
+        self._mapping = statedict["_mapping"]
+
+    def __iter__(self) -> Iterator[Trial]:
+        return iter(self.original_registry)
+
+    def __len__(self) -> int:
+        return len(self.original_registry)
+
+    def __contains__(self, trial: Trial):
+        return trial in self.original_registry
+
+    def __getitem__(self, item: Trial) -> list[Trial]:
+        trial_id = _get_id(item)
+        if trial_id not in self._mapping:
+            raise KeyError(item)
+        transformed_trial_ids = self._mapping[trial_id]
+        return [
+            self.transformed_registry[transformed_id]
+            for transformed_id in transformed_trial_ids
+        ]
+
+    def get_trials(self, original_trial: Trial) -> list[Trial]:
+        """Return the registered transformed trials that map to the given trial in the original
+        space.
+        """
+        return self.get(original_trial, [])
+
+    def register(self, original_trial: Trial, transformed_trial: Trial) -> str:
+        """Register an equivalence between the given original trial and the transformed trial."""
+        # NOTE: Choosing not to register the trials here, and instead do it more manually.
+        # original_id = self.original_registry.register(original_trial)
```

Comment:
On one hand it may be confusing to have the side effect of populating the registries when calling mapping.register. On the other hand we risk having incoherent original and transformed registries with the mapping if we rely on updating them externally. Handling them here would not block external updates on the registries anyhow, so maybe we are at risk for incoherence anyway and we better make it explicit like you do here. :thinking:

Reply:
Sorry to ask, but do I need to change something here?

Reply:
No I'm sorry :smile: I was just sharing my thoughts in case we should discuss this, but I latter agreed with your solution: https://github.com/Epistimio/orion/pull/833#discussion_r834688021